File: app/src/main/java/com/themovielist/moviedetail/MovieDetailContract.java
package com.themovielist.moviedetail;

import com.themovielist.base.BasePresenter;
import com.themovielist.model.MovieModel;
import com.themovielist.model.MovieReviewModel;
import com.themovielist.model.MovieTrailerModel;

import java.util.List;

/**
 * Presenter of the Movie Detail Fragment.
 */

public abstract class MovieDetailContract {
    public interface View {
        void showMovieReview(List<MovieReviewModel> movieReviewModelList);

        void showMovieDetail(MovieModel movieModel);

        void showMovieTrailer(List<MovieTrailerModel> movieTrailerList);

        void setFavoriteButtonState(boolean favorite);

        void showSuccessMessageAddFavoriteMovie();

        void showSuccessMessageRemoveFavoriteMovie();


        void showErrorMessageAddFavoriteMovie();

        void showErrorMessageRemoveFavoriteMovie();

        void showErrorMessageLoadReviews();

        void showErrorMessageLoadTrailers();

        void setShowAllReviewsButtonVisibility(boolean visible);

        void setShowAllTrailersButtonVisibility(boolean visible);

        void showLoadingReviewsIndicator();

        void showLoadingTrailersIndicator();

        void showAllReviews(List<MovieReviewModel> movieReviewList, boolean hasMore);

        void showAllTrailers(List<MovieTrailerModel> movieTrailerList);

        void showEmptyReviewListMessage();

        void showEmptyTrailerListMessage();
    }

    public interface Presenter extends BasePresenter<View> {
        void start(MovieModel movieModel);

        void removeFavoriteMovie(MovieModel movieModel);

        void saveFavoriteMovie(MovieModel movieModel);

        void showAllReviews();

        void showAllTrailers();

        void tryToLoadTrailersAgain();

        void tryToLoadReviewAgain();
    }
}

File: app/src/main/java/com/themovielist/moviedetail/MovieDetailContract.kt
package com.themovielist.moviedetail

import com.themovielist.base.BasePresenter
import com.themovielist.model.MovieModel
import com.themovielist.model.MovieReviewModel
import com.themovielist.model.MovieTrailerModel
import com.themovielist.model.MovieWithGenreModel

/**
 * Presenter of the Movie Detail Fragment.
 */

abstract class MovieDetailContract {
    interface View {
        fun showMovieInfo(movieWithGenreModel: MovieWithGenreModel)

        fun setFavoriteButtonState(favorite: Boolean)

        fun showSuccessMessageAddFavoriteMovie()

        fun showSuccessMessageRemoveFavoriteMovie()

        fun showErrorMessageAddFavoriteMovie()

        fun showErrorMessageRemoveFavoriteMovie()

        fun showAllReviews(movieReviewList: List<MovieReviewModel>, hasMore: Boolean)

        fun showAllTrailers(movieTrailerList: List<MovieTrailerModel>)

        fun setFavoriteButtonEnabled(enabled: Boolean)
        fun showLoadingMovieDetailIndicator()
        fun hideLoadingMovieDetailIndicator()
        fun showErrorLoadingMovieDetail(error: Throwable)
        fun showMovieDetailInfo()
        fun showMovieRuntime(hourMinute: Pair<Int, Int>)
        fun bindMovieReviewInfo(movieReviewList: List<MovieReviewModel>)
        fun dispatchFavoriteMovieEvent(movieModel: MovieModel, isFavorite: Boolean)
        fun bindMovieTrailerInfo(movieTrailerList: List<MovieTrailerModel>)
    }

    interface Presenter : BasePresenter<View> {
        fun start(movieModel: MovieModel)

        fun showAllReviews()

        fun showAllTrailers()
        fun tryFecthMovieDetailAgain()
    }
}
